contracts/token/ERC721/extensions/draft-ERC721Royalty.sol

```diff
@@ -0,0 +1,125 @@
+    mapping(uint256 => RoyaltyInfo) private _tokenRoyaltyInfo;
+
+    /**
+     * @dev Sets tokens royalties
+     *
+     * Requirements:
+     * - `tokenId` must be already mined.
+     * - `receiver` cannot be the zero address.
+     * - `fraction` must indicate the percentage fraction using two decimals.
+     */
+    function _setTokenRoyalty(
+        uint256 tokenId,
+        address receiver,
+        uint256 fraction
+    ) internal virtual {
+        require(fraction > 0, "ERC2981: Royalty percentage is too low");
+        require(fraction <= _feeDenominator(), "ERC2981: Royalty percentage will exceed salePrice");
+        require(receiver != address(0), "ERC2981: Invalid receiver");
+        require(_exists(tokenId), "ERC2981: Nonexistent token");
+
+        _tokenRoyaltyInfo[tokenId] = RoyaltyInfo(receiver, fraction);
+    }
+
+    /**
+     *
+     * @dev Sets global royalty
+     *
+     * Requirements:
+     * - `receiver` cannot be the zero address.
+     * - `fraction` must indicate the percentage fraction. Needs to be set appropriately
+     * according to the _feeDenominator granularity.
+     */
+    function _setGlobalRoyalty(address receiver, uint256 fraction) internal virtual {
+        require(fraction > 0, "ERC2981: Royalty percentage is too low");
+        require(fraction <= _feeDenominator(), "ERC2981: Royalty percentage will exceed salePrice");
+        require(receiver != address(0), "ERC2981: Invalid receiver");
+
+        _globalRoyaltyInfo = RoyaltyInfo(receiver, fraction);
+    }
+
+    /**
+     * @dev See {IERC721Royalty-royaltyInfo}
+     */
+    function royaltyInfo(uint256 _tokenId, uint256 _salePrice) external view override returns (address, uint256) {
+        RoyaltyInfo memory royalty;
+
+        if (_tokenRoyaltyInfo[_tokenId].receiver != address(0)) {
+            royalty = _tokenRoyaltyInfo[_tokenId];
+        } else {
+            royalty = _globalRoyaltyInfo;
+        }
+
+        uint256 royaltyAmount = (_salePrice * royalty.royaltyFraction) / _feeDenominator();
+
+        return (royalty.receiver, royaltyAmount);
+    }
+
+    /**
+     * @dev See {IERC165-supportsInterface}.
+     */
+    function supportsInterface(bytes4 interfaceId) public view virtual override(ERC721, IERC165) returns (bool) {
+        return interfaceId == type(IERC721Royalty).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev Returns the percentage granularity being used. The default denominator is 10000
+     *  but it can be customized by an override.
+     */
+    function _feeDenominator() internal pure virtual returns (uint256) {
+        return 10000;
+    }
+
+    /**
+     * @dev Removes `tokenId` royalty information.
+     * The royalty information is cleared when the token is burned.
+     *
+     * Requirements:
+     *
+     * - `tokenId` royalty information must exist.
+     *
+     */
+    function _deleteTokenRoyalty(uint256 tokenId) internal virtual {
```

Comment:
there is no version of that for the global ?

Reply:
You are right, is missing, I'll add it

Reply:
Note that if you remove the checks `fraction > 0` and `recipient != address(0)` this could possibly not be needed

maybe these two checks can be merged into 
```
require(recipient != address(0) || fraction == 0);
```

Reply:
Wait upon further thinking it should be or and not xor.

We want to let token have a royalty fraction of 0 with a recipient that is not 0 → this is how you enforce that a token has fraction 0 when the global is not 0

So
```
require(recipient != address(0) || fraction == 0);
```

Reply:
I've also included a new test case for this scenario